File: algorithms/algorithms-cplusplus/leetcode/Question_0098_Validate_Binary_Search_Tree.cc
#include <stack>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
 public:
  bool isValidBST(TreeNode* root) {
    stack<TreeNode*> st;
    TreeNode *pre = NULL;
    while (root || !st.empty()) {
      if (root) {
        st.push(root);
        root = root->left;
      } else {
        root = st.top();
        st.pop();
        if (pre && pre->val >= root->val) {
          return false;
        }
        pre = root;
        root = root->right;
      }
    }
    return true;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0098_Validate_Binary_Search_Tree.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.Stack;

public class Question_0098_Validate_Binary_Search_Tree {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public boolean isValidBST(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        TreeNode pre = null;
        while (root != null || !stack.empty()) {
            if (root != null) {
                stack.push(root);
                root = root.left;
            } else {
                root = stack.pop();
                if (pre != null && pre.val >= root.val) {
                    return false;
                }
                pre = root;
                root = root.right;
            }
        }
        return true;
    }
}
